This is a guided project I have learned recently on Dataquest. Data Analysis by Python. This project focuses on Numpy and Pandas. In this guided project, I'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.The aim of this project is to clean the data and analyze the included used car listings.Furthermore, I am going to be familiar with Numpy and Pandas.

# Exploring Ebay car sale data

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

From the work above, we can make the following observations:
- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores. 

I'll convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

# Cleaning column names
Use the DataFrame.columns to change some name of columns:
- yearOfRegistration to registration_year
- monthOfRegistration to registration_month
- notRepairedDamage to unrepaired_damage
- dateCreated to ad_created
The rest of the columnn names from camelCase to snake_case.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos_test = autos.copy()
autos_test_column = autos_test.columns

In [7]:
#convert camelCase to snake_case
def change_case(text):
    res = [text[0].lower()]
    for c in text[1:]:
        if c in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'):
            res.append('_')
            res.append(c.lower())
        else:
            res.append(c)
    return ''.join(res)

In [8]:
new_columns = []
for name in autos_test.columns:
    new_name = change_case(name)
    new_columns.append(new_name)
new_columns[7] = "registration_year"
new_columns[9] = "powerPS"
new_columns[12] = "registration_month"
new_columns[14] = "unrepaired_damage"
new_columns[15] = "ad_created"
autos.columns = new_columns

In [9]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'unrepaired_damage',
       'ad_created', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


# Innitial exploration and cleaning

In [10]:
autos.describe()

price  registration_year        powerPS      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count       371528.000000        371528.0  371528.00000  
mean             5.734445             0.0   50820.66764  
std              3.712412             0.0   25799.08247  
min              0.000000             0.0    1067.00000  
25%              3.000000             0.0   30459.00000  
50%              6.000000             0.0   49610.00000  
75%              9.000000             0.0   71546.00000  
max             12.000000             0.0   99998.00000

From the work above:
- any columns that have mostly one value that are candidates to be dropped
- Any columns that need more investigation.
- Price column and kilometer column are numeric data stored as text that needs to be cleaned.

Change Price column and Kilometer column to numeric data by:
+ Remove any non-numeric characters
+ Convert the column to a numeric dtype
+ Rename the column to odometer_km

In [11]:
price_unique = autos["price"].unique()
price_unique

array([  480, 18300,  9800, ..., 18429, 24895, 10985])

In [12]:
odometer_unique = autos["kilometer"].unique()
odometer_unique

array([150000, 125000,  90000,  40000,  30000,  70000,   5000, 100000,
        60000,  20000,  80000,  50000,  10000])

In [13]:
autos["vehicle_type"].dtypes

dtype('O')

# Exploring the date columns

In [14]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled ad_created            last_seen
0  2016-03-24 11:52:17        NaN  2016-04-07 03:16:57
1  2016-03-24 10:58:45         ja  2016-04-07 01:46:50
2  2016-03-14 12:52:21        NaN  2016-04-05 12:47:46
3  2016-03-17 16:54:04       nein  2016-03-17 17:40:17
4  2016-03-31 17:25:20       nein  2016-04-06 10:17:21

In [15]:
print(autos['date_crawled'].str[:10])

0         2016-03-24
1         2016-03-24
2         2016-03-14
3         2016-03-17
4         2016-03-31
             ...    
371523    2016-03-14
371524    2016-03-05
371525    2016-03-19
371526    2016-03-20
371527    2016-03-07
Name: date_crawled, Length: 371528, dtype: object


In [16]:
autos['date_crawled'].value_counts(normalize=True, dropna=False)

2016-03-24 14:49:47    0.000019
2016-03-19 21:49:56    0.000016
2016-03-26 22:57:31    0.000016
2016-03-22 10:42:10    0.000013
2016-03-23 13:50:19    0.000013
                         ...   
2016-03-05 15:06:22    0.000003
2016-03-19 17:55:38    0.000003
2016-03-23 19:25:25    0.000003
2016-03-22 00:32:40    0.000003
2016-03-13 21:38:43    0.000003
Name: date_crawled, Length: 280500, dtype: float64

In [17]:
autos['date_crawled'].sort_index()

0         2016-03-24 11:52:17
1         2016-03-24 10:58:45
2         2016-03-14 12:52:21
3         2016-03-17 16:54:04
4         2016-03-31 17:25:20
                 ...         
371523    2016-03-14 17:48:27
371524    2016-03-05 19:56:21
371525    2016-03-19 18:57:12
371526    2016-03-20 19:41:08
371527    2016-03-07 19:39:19
Name: date_crawled, Length: 371528, dtype: object

In [18]:
autos['date_crawled'].value_counts()

2016-03-24 14:49:47    7
2016-03-19 21:49:56    6
2016-03-26 22:57:31    6
2016-03-22 10:42:10    5
2016-03-23 13:50:19    5
                      ..
2016-03-05 15:06:22    1
2016-03-19 17:55:38    1
2016-03-23 19:25:25    1
2016-03-22 00:32:40    1
2016-03-13 21:38:43    1
Name: date_crawled, Length: 280500, dtype: int64

## Dealing with registration year

In [19]:
autos['registration_year'].describe()

count    371528.000000
mean       2004.577997
std          92.866598
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

As I see, the maximum year is 9999 - many years into the future, the minimum year is 1000 - before cars are invented. So I want to remove car listings with year_registration outside the 1900 - 2016.

In [20]:
filtered_autos= autos[(autos["registration_year"] > 1900) & (autos["registration_year"] < 2016)]

In [21]:
filtered_autos.head(20)

date_crawled                                               name  \
0   2016-03-24 11:52:17                                         Golf_3_1.6   
1   2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
2   2016-03-14 12:52:21                     Jeep_Grand_Cherokee_"Overland"   
3   2016-03-17 16:54:04                                 GOLF_4_1_4__3TÜRER   
4   2016-03-31 17:25:20                     Skoda_Fabia_1.4_TDI_PD_Classic   
5   2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6   2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
7   2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
8   2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   
9   2016-03-17 10:53:50  VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...   
10  2016-03-26 19:54:18                                  Mazda_3_1.6_Sport   
11  2016-04-07 10:06:22      Volkswagen_Passat_Variant_2.0_TDI_Comfortline   
12  2016-03-15 22:49:09                  VW_Passat_Facelift_35i__"7Sitzer"   
13  2016-03-21 21:37:40                     VW_PASSAT_1.9_TDI_131_PS_LEDER   
14  2016-03-21 12:57:01  Nissan_Navara_2.5DPF_SE4x4_Klima_Sitzheizg_Blu...   
15  2016-03-11 21:39:15                       KA_Lufthansa_Edition_450€_VB   
17  2016-03-20 10:25:19                        Renault_Twingo_1.2_16V_Aut.   
18  2016-03-23 15:48:05                   Ford_C_MAX_2.0_TDCi_DPF_Titanium   
19  2016-04-01 22:55:47                  Mercedes_Benz_A_160_Classic_Klima   
20  2016-04-01 19:56:48                  Volkswagen_Scirocco_1.4_TSI_Sport   

    seller offer_type  price   abtest vehicle_type  registration_year  \
0   privat    Angebot    480     test          NaN               1993   
1   privat    Angebot  18300     test        coupe               2011   
2   privat    Angebot   9800     test          suv               2004   
3   privat    Angebot   1500     test   kleinwagen               2001   
4   privat    Angebot   3600     test   kleinwagen               2008   
5   privat    Angebot    650     test    limousine               1995   
6   privat    Angebot   2200     test       cabrio               2004   
7   privat    Angebot      0     test    limousine               1980   
8   privat    Angebot  14500  control          bus               2014   
9   privat    Angebot    999     test   kleinwagen               1998   
10  privat    Angebot   2000  control    limousine               2004   
11  privat    Angebot   2799  control        kombi               2005   
12  privat    Angebot    999  control        kombi               1995   
13  privat    Angebot   2500  control        kombi               2004   
14  privat    Angebot  17999  control          suv               2011   
15  privat    Angebot    450     test   kleinwagen               1910   
17  privat    Angebot   1750  control   kleinwagen               2004   
18  privat    Angebot   7550     test          bus               2007   
19  privat    Angebot   1850     test          bus               2004   
20  privat    Angebot  10400  control        coupe               2009   

      gearbox  powerPS     model  kilometer  registration_month fuel_type  \
0     manuell        0      golf     150000                   0    benzin   
1     manuell      190       NaN     125000                   5    diesel   
2   automatik      163     grand     125000                   8    diesel   
3     manuell       75      golf     150000                   6    benzin   
4     manuell       69     fabia      90000                   7    diesel   
5     manuell      102       3er     150000                  10    benzin   
6     manuell      109   2_reihe     150000                   8    benzin   
7     manuell       50    andere      40000                   7    benzin   
8     manuell      125     c_max      30000                   8    benzin   
9     manuell      101      golf     150000                   0     

In [22]:
filtered_autos['registration_year'].value_counts(normalize = True)

2000    0.070768
1999    0.065626
2005    0.064326
2006    0.058313
2001    0.058278
          ...   
1925    0.000003
1920    0.000003
1919    0.000003
1915    0.000003
1911    0.000003
Name: registration_year, Length: 96, dtype: float64

In [23]:
filtered_autos.shape

(346921, 20)

In [24]:
filtered_autos['unrepaired_damage'].value_counts(dropna=False)

volkswagen        73259
bmw               38489
opel              36716
mercedes_benz     33614
audi              31273
ford              23720
renault           16275
peugeot           10357
fiat               8880
seat               6415
skoda              5400
mazda              5303
smart              4915
citroen            4810
nissan             4712
toyota             4462
sonstige_autos     3728
hyundai            3441
mini               3245
volvo              3208
mitsubishi         2854
honda              2616
kia                2404
alfa_romeo         2203
suzuki             2181
porsche            2148
chevrolet          1763
chrysler           1377
dacia               860
jeep                773
daihatsu            751
land_rover          749
subaru              732
jaguar              612
trabant             572
saab                514
daewoo              494
lancia              445
rover               442
lada                209
Name: unrepaired_damage, dtype: int64

## Exploring price by brand

I want to find out the average price of top brand car in this dataset and storing aggregate in dataframe

In [25]:
# Lọc ra top 20 brand xe bán chạy

# Tính giá trung bình của brand xe có giá trị cao
# append vào bảng dataframe

In [26]:
# Loc ra top 20 brand xe ban chay
top_20 = filtered_autos['unrepaired_damage'].value_counts(dropna=False).head(20)
top_20

volkswagen        73259
bmw               38489
opel              36716
mercedes_benz     33614
audi              31273
ford              23720
renault           16275
peugeot           10357
fiat               8880
seat               6415
skoda              5400
mazda              5303
smart              4915
citroen            4810
nissan             4712
toyota             4462
sonstige_autos     3728
hyundai            3441
mini               3245
volvo              3208
Name: unrepaired_damage, dtype: int64

In [27]:
#Tinh gia trung binh cua nhung brand co gia tri cao
top_20.index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'smart', 'citroen',
       'nissan', 'toyota', 'sonstige_autos', 'hyundai', 'mini', 'volvo'],
      dtype='object')

In [28]:
top20 = list(top_20.index)

In [29]:
avg_price_name = {}
for name in top20:
    mask = filtered_autos.unrepaired_damage == name
    filtered_name_car = filtered_autos[mask]
    avg_price = filtered_name_car['price'].mean()
    avg_price_name[name]=avg_price

In [30]:
avg_price_name

{'volkswagen': 14023.393275911492,
 'bmw': 14957.060692665436,
 'opel': 3316.4628499836585,
 'mercedes_benz': 17850.10397453442,
 'audi': 16425.188245451347,
 'ford': 8936.607124789207,
 'renault': 2409.466728110599,
 'peugeot': 3243.0004827652797,
 'fiat': 4239.923873873874,
 'seat': 4467.764146531566,
 'skoda': 6480.099814814815,
 'mazda': 5919.853102017726,
 'smart': 3606.852899287894,
 'citroen': 9374.67318087318,
 'nissan': 4643.384762308999,
 'toyota': 5288.300089645899,
 'sonstige_autos': 653067.4737124464,
 'hyundai': 5531.001453065969,
 'mini': 9977.885053929122,
 'volvo': 5133.542705735661}

Sonstige_autos car is too expensive.

In [31]:
mask_sonstige_autos =filtered_autos.unrepaired_damage == 'sonstige_autos'
abc = filtered_autos[mask_sonstige_autos]

In [32]:
abc.price.describe()

count    3.728000e+03
mean     6.530675e+05
std      3.524741e+07
min      0.000000e+00
25%      8.990000e+02
50%      4.825000e+03
75%      1.370000e+04
max      2.147484e+09
Name: price, dtype: float64

## Storing to Dataframe

In [33]:
pd.DataFrame.from_dict(avg_price_name, orient='index',columns = ['avg_price'])

avg_price
volkswagen       14023.393276
bmw              14957.060693
opel              3316.462850
mercedes_benz    17850.103975
audi             16425.188245
ford              8936.607125
renault           2409.466728
peugeot           3243.000483
fiat              4239.923874
seat              4467.764147
skoda             6480.099815
mazda             5919.853102
smart             3606.852899
citroen           9374.673181
nissan            4643.384762
toyota            5288.300090
sonstige_autos  653067.473712
hyundai           5531.001453
mini              9977.885054
volvo             5133.542706